File: year2024/day01/typescript/solution.ts
import { readFileSync } from 'fs';

const file: any = readFileSync('../' + (process.argv[2] === 'puzzle' ? 'puzzle' : 'example') + '.txt', 'utf-8');
const input = file.toString().split('\n').reduce((a, l) => (l.split(/\s+/).map(Number).forEach((n, i) => a[i].push(n)), a), [[], []]).map(arr => arr.sort((x, y) => x - y));
if(process.argv[2] === 'puzzle') {
  console.log(part_1());
  console.log(part_2());
}

export function part_1(): number {
  return input[0].reduce((a, v, i) => a + Math.abs(v - input[1][i]), 0);
}

export function part_2(): number {
  return input[0].reduce((a, v) => a + v * input[1].filter(n => n === v).length, 0);
}

File: year2024/day01/python/solution.py
def main():
  print(part_1())
  print(part_2())


def part_1():
  return sum(abs(v - input[1][i]) for i, v in enumerate(input[0]))


def part_2():
  return sum(v * input[1].count(v) for v in input[0])


def setup(path):
    global input

    with open(path, 'r') as file:
        lines = file.read().strip().split('\n')
    
    input = [[], []]
    for line in lines:
      numbers = list(map(int, line.split()))
      for i, num in enumerate(numbers):
        input[i].append(num)

    input[:] = [sorted(arr) for arr in input]


if __name__ == '__main__':
  setup('../puzzle.txt')
  main()